pkg/controller/direct/eventarc/googlechannelconfig_controller.go

```diff
@@ -0,0 +1,218 @@
+}
+
+func (m *googleChannelConfigModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.EventarcGoogleChannelConfig{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewGoogleChannelConfigIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get eventarc GCP client
+	gcpClient, err := newGCPClient(ctx, &m.config)
+	if err != nil {
+		return nil, err
+	}
+	eventarcClient, err := gcpClient.newEventarcClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &googleChannelConfigAdapter{
+		gcpClient: eventarcClient,
+		id:        id,
+		desired:   obj,
+		reader:    reader,
+	}, nil
+}
+
+func (m *googleChannelConfigModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type googleChannelConfigAdapter struct {
+	gcpClient *gcp.Client
+	id        *krm.GoogleChannelConfigIdentity
+	desired   *krm.EventarcGoogleChannelConfig
+	actual    *pb.GoogleChannelConfig
+	reader    client.Reader
+}
+
+var _ directbase.Adapter = &googleChannelConfigAdapter{}
+
+func (a *googleChannelConfigAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting eventarc googlechannelconfig", "name", a.id)
+
+	req := &pb.GetGoogleChannelConfigRequest{Name: a.id.String()}
+	actual, err := a.gcpClient.GetGoogleChannelConfig(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting eventarc googlechannelconfig %q: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *googleChannelConfigAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	// GoogleChannelConfig is a singleton resource that cannot be created.
+	// It can only be fetched and updated.
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating eventarc googlechannelconfig is a no-op", "name", a.id)
+	return nil
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *googleChannelConfigAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating eventarc googlechannelconfig", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	desired := a.desired.DeepCopy()
+	resource := EventarcGoogleChannelConfigSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	paths := []string{}
+	paths = append(paths, "crypto_key_name")
```

Comment:
I don't think we should hardcode this field to be always updated.

Reply:
Update has only one field. How do we deal with this ?

Reply:
More so with removing the reference is an update. removing key ref is the only update.

Reply:
I think we need to compare desired vs. actual and skip the update if there is no diff in this field?

Reply:
done.